service/storage_service.cc

```diff
@@ -1876,33 +1876,56 @@ future<std::map<gms::inet_address, float>> storage_service::effective_ownership(
             }
             keyspace_name = "system_traces";
         }
-        auto token_ownership = dht::token::describe_ownership(ss._token_metadata.sorted_tokens());
-
-        std::map<gms::inet_address, float> final_ownership;
-
-        // calculate ownership per dc
-        for (auto endpoints : ss._token_metadata.get_topology().get_datacenter_endpoints()) {
-            // calculate the ownership with replication and add the endpoint to the final ownership map
-            for (const gms::inet_address& endpoint : endpoints.second) {
-                float ownership = 0.0f;
-                for (range<token> r : ss.get_ranges_for_endpoint(keyspace_name, endpoint)) {
-                    // get_ranges_for_endpoint will unwrap the first range.
-                    // With t0 t1 t2 t3, the first range (t3,t0] will be splitted
-                    // as (min,t0] and (t3,max]. Skippping the range (t3,max]
-                    // we will get the correct ownership number as if the first
-                    // range were not splitted.
-                    if (!r.end()) {
-                        continue;
-                    }
-                    auto end_token = r.end()->value();
-                    if (token_ownership.find(end_token) != token_ownership.end()) {
-                        ownership += token_ownership[end_token];
-                    }
-                }
-                final_ownership[endpoint] = ownership;
-            }
-        }
-        return final_ownership;
+
+        auto final_ownership = make_lw_shared<std::map<gms::inet_address, float>>();
+        // The following loops seems computationally heavy, but it's not as bad.
+        // The upper two simply iterate over all the endpoints by iterating over all the
+        // DC and all the instances in each DC.
+        //
+        // The call for get_range_for_endpoint is done once per endpoint
+        return do_with(std::map<token, float>(dht::token::describe_ownership(ss._token_metadata.sorted_tokens())),
```

Comment:
Why the extra constructor call? describe_ownership already returns a map<token, float>.

Reply:
I'll add a move

Reply:
Why?

Reply:
To remove the constructor call

Reply:
It's enough to remove it. No need for an extra move.

Reply:
if you are ok with not explicitly stating the type, great

Reply:
I don't understand.

Reply:
I meant that type will be deduce and will not be stated explicitly.

Reply:
The type is explicit in the argument list of the lambda:

```c++
        [&ss, keyspace_name, final_ownership](const std::map<token, float>& token_ownership, 
```